doc/architecture.md

```diff
@@ -0,0 +1,206 @@
+  `DandiDav` instance in order to fetch information about the specified
+  resource(s).
+
+
+`DandiClient`
+-------------
+
+- The [`DandiClient`][] type is used to retrieve information about resources on
+  an Archive instance (the one specified via the `--api-url` option on program
+  invocation) via a combination of the Archive's API and (for entries & folders
+  inside Zarrs) requests to an S3 bucket.  It is the data source for the
+  `/dandisets/` hierarchy served by `dandidav`.
+
+    - Information about child resources of Zarrs needs to be fetched via S3
+      directly rather than using the Archive API because the latter currently
+      does not provide an efficient way to list folders within a Zarr.
+
+- The `DandiClient` API is modeled somewhat after
+  [`octocrab`](https://github.com/XAMPPRocky/octocrab): To operate on an API
+  resource, an endpoint object is requested from the `DandiClient` (possibly
+  indirectly via a super-endpoint), and the endpoint's methods are called to
+  obtain one or more objects that provide information about the resource or its
+  children, but these resource objects are distinct from the endpoint objects
+  and do not have any methods for making API calls themselves.
+
+- The code assumes Zarr assets have a `contentUrl` that points to the Zarr's
+  entries on an S3 bucket; specifically, the first element of `contentUrl` that
+  can be parsed by [`S3Location::parse_url()`][s3loc-parse] into a bucket,
+  optional region, and key prefix is used as such.  It is then assumed that the
+  Zarr's entries are laid out under the given key prefix (after appending a
+  trailing slash if one is not already present) on the given bucket with the
+  same names & directory structure as the actual Zarr.
+
+    - If a given Zarr asset lacks an S3 `contentUrl`, any requests to a path
+      under that Zarr will result in a 500 response.
+
+    - If an S3 URL does not specify the bucket region, it is determined via a
+      `HEAD` request to the domain.  The S3 client caching (see below) means
+      that this will almost certainly be done at most once per `dandidav`
+      process.
+
+    - The S3 client used to fetch information about Zarr entries from a given
+      bucket is cached in the `DandiClient`.  While it is almost certain that
+      all Zarrs in the same Archive instance will use the same bucket and thus
+      the same client, the possibility of multiple buckets is guarded against
+      by using a cache of up to [`S3CLIENT_CACHE_SIZE`][] clients.
+
+
+`ZarrManClient`
+---------------
+
+The [`ZarrManClient`][] type is used to retrieve information about Zarr entries
+by fetching Zarr manifests from <https://github.com/dandi/zarr-manifests> via
+the mirror at <https://datasets.datalad.org/dandi/zarr-manifests/>.  It is the
+data source for the `/zarrs/` hierarchy served by `dandidav`.
+
+See the documentation for the `dandidav::zarrman` module for more information.
```

Comment:
it is not really descriptive there ATM. Would be nice to get (follow up PR) some brief similar description for operations on zarrs.

Reply:
Expanded in bb682a1.